File: client-mobile/IntegratedLife/Models/Health/Workout.swift
import Foundation

struct WorkoutExercise: Codable, Identifiable {
	let exerciseId: String
	let name: String?
	let order: Int
	let defaultSets: Int?
	let defaultReps: Int?
	let defaultWeight: Double?

	var id: String { exerciseId }
}

struct RecurrenceRule: Codable {
	var frequency: String
	var interval: Int
	var daysOfWeek: [String]?
	var dayOfMonth: Int?
	var endDate: String?
	var count: Int?
}

struct Workout: Codable, Identifiable {
	let id: String
	let name: String
	let difficulty: String
	let isGlobal: Bool
	let userId: String?
	let exercises: [WorkoutExercise]?
	let exerciseCount: Int?
	let schedule: RecurrenceRule?
}

struct CreateWorkoutRequest: Codable {
	let name: String
	let difficulty: String
	let exercises: [WorkoutExerciseInput]
	let schedule: RecurrenceRule?
}

struct WorkoutExerciseInput: Codable {
	let exerciseId: String
	let order: Int
	let defaultSets: Int?
	let defaultReps: Int?
	let defaultWeight: Double?
}

File: server/src/models/Workout.ts
import mongoose, { Schema, Document, Model } from 'mongoose'

export type WorkoutExerciseSubdoc = {
	exerciseId: mongoose.Types.ObjectId
	order: number
	defaultSets?: number
	defaultReps?: number
	defaultWeight?: number
}

export type RecurrenceRuleSubdoc = {
	frequency: string
	interval: number
	daysOfWeek?: string[]
	dayOfMonth?: number
	endDate?: Date
	count?: number
}

export type WorkoutDocument = Document & {
	name: string
	difficulty: string
	isGlobal: boolean
	userId?: mongoose.Types.ObjectId
	exercises: WorkoutExerciseSubdoc[]
	schedule?: RecurrenceRuleSubdoc
	createdAt: Date
	updatedAt: Date
}

const workoutExerciseSchema = new Schema<WorkoutExerciseSubdoc>(
	{
		exerciseId: { type: Schema.Types.ObjectId, ref: 'Exercise', required: true },
		order: { type: Number, required: true },
		defaultSets: { type: Number },
		defaultReps: { type: Number },
		defaultWeight: { type: Number }
	},
	{ _id: false }
)

const recurrenceRuleSchema = new Schema<RecurrenceRuleSubdoc>(
	{
		frequency: { type: String, required: true },
		interval: { type: Number, default: 1 },
		daysOfWeek: [{ type: String }],
		dayOfMonth: { type: Number },
		endDate: { type: Date },
		count: { type: Number }
	},
	{ _id: false }
)

const workoutSchema = new Schema<WorkoutDocument>(
	{
		name: { type: String, required: true },
		difficulty: { type: String, required: true },
		isGlobal: { type: Boolean, default: false },
		userId: { type: Schema.Types.ObjectId, ref: 'User' },
		exercises: [workoutExerciseSchema],
		schedule: { type: recurrenceRuleSchema }
	},
	{ timestamps: true }
)

workoutSchema.index({ userId: 1 })
workoutSchema.index({ isGlobal: 1 })
workoutSchema.index({ difficulty: 1 })

export const Workout: Model<WorkoutDocument> =
	mongoose.models.Workout ?? mongoose.model<WorkoutDocument>('Workout', workoutSchema)
